### PR TITLE
[FIRRTL] Update Subfield op string fieldName to integer fieldIndex

### PR DESCRIPTION
**Pull request by prithayan**
In the FIRRTL dialect, the `bundle` elements are accessed using the `subfield` operation with a string field name. 
Each of the element also has a corresponding index in the `bundle`.  
This PR replaces the string field of a `subfield` with the corresponding index of the element in the `bundle`.
For example, replace : 
``%0 = firrtl.subfield %_M_read("data") : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: sint<42>>) -> !firrtl.sint<42>``
with ==> 
``%0 = firrtl.subfield %_M_read(3) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: sint<42>>) -> !firrtl.sint<42>``

This change should avoid a lot of string comparisons when handling `SubfieldOp`.

### CODE REVIEW

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 1]
```diff
@@ -543,9 +543,9 @@ static ValueVectorList extractSubfields(FModuleOp subModuleOp,
     auto type = arg.getType().cast<FIRRTLType>();
     if (auto bundleType = type.dyn_cast<BundleType>()) {
       // Extract all subfields of all bundle ports.
-      for (auto &element : bundleType.getElements()) {
-        subfields.push_back(
-            rewriter.create<SubfieldOp>(insertLoc, arg, element.name));
+      for (size_t index = 0, size = bundleType.getNumElements(); index < size;
+           ++index) {
+        subfields.push_back(rewriter.create<SubfieldOp>(insertLoc, arg, index));
```
Comment (seldridge) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:548]: nit: using shorter variable names can save a line:
```suggestion
      for (size_t i = 0, e = bundleType.getNumElements(); i != e; ++i) {
        subfields.push_back(rewriter.create<SubfieldOp>(insertLoc, arg, i));
```

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 2]
```diff
@@ -1714,13 +1714,16 @@ bool HandshakeBuilder::visitHandshake(MemoryOp op) {
     // Create a subfield op to access this port in the memory.
     auto fieldName = loadIdentifier(i);
     auto memBundle = memOp.getPortNamed(fieldName);
+    auto memType = memBundle.getType().cast<FIRRTLType>().cast<BundleType>();
```
Comment (seldridge) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1717]: Double cast?
```suggestion
    auto memType = memBundle.getType().cast<BundleType>();
```

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 3]
```diff
@@ -1519,40 +1519,32 @@ void ConstantOp::build(OpBuilder &builder, OperationState &result,
...
+      getAttr<IntegerAttr>(attrs, "fieldIndex").getValue().getZExtValue();
 
-  if (loc)
-    mlir::emitError(*loc, "subfield requires bundle operand");
-  return {};
+  // SubfieldOp verifier checks that the field index is valid with number of
+  // subelements. FIRRTL puts flips on element fields, not on the underlying
+  // types.  The result type of a subfield should strip a flip
+  // if one exists.
+  return inType.cast<BundleType>().getElement(fieldIndex).getValue().type;
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1541]: This comment about stripping flips is stale. I think you can just strike the last two sentences.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 4]
```diff
@@ -1513,8 +1513,9 @@ void FIRStmtParser::emitInvalidate(Value val, Flow flow) {
   // to only the leaf sources.
   TypeSwitch<FIRRTLType>(tpe)
       .Case<BundleType>([&](auto tpe) {
-        for (auto element : tpe.getElements()) {
-          auto subfield = builder.create<SubfieldOp>(val, element.name);
+        for (size_t index = 0, size = tpe.getNumElements(); index < size;
+             ++index) {
+          auto subfield = builder.create<SubfieldOp>(val, index);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:1518]: Nit: can save a line:
```suggestion
        for (size_t i = 0, e = tpe.getNumElements(); i != e; ++i) {
          auto subfield = builder.create<SubfieldOp>(val, i);
```

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 5]
```diff
@@ -1665,11 +1666,18 @@ ParseResult FIRStmtParser::parsePostFixFieldId(Value &result) {
   StringRef fieldName;
   if (parseFieldId(fieldName, "expected field name"))
     return failure();
-
+  auto indexV =
+      result.getType().cast<FIRRTLType>().cast<BundleType>().getElementIndex(
+          fieldName);
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:1671]: Double cast shouldn't be necessary?
```suggestion
      result.getType().cast<BundleType>().getElementIndex(fieldName);
```

**Review by seldridge — APPROVED** [review]
Overall, this looks like a great improvement. I would expect that this will save on the overall memory footprint a fair amount?

Some nitty comments throughout.

I was slightly concerned about the readability of the printed MLIR text. However, I'm less so after looking at the test cases. (It's pretty easy to eyeball what's going on. Large bundles are still a problem, but they were arguably a problem even when using strings.)

**Comment by prithayan** [comment]
Thanks @seldridge , updated with the suggested changes. 
Updating the parser and printer for printing a more readable IR with string field names, is also an option we can consider.

**Review by seldridge — APPROVED** [review]

File: include/circt/Dialect/FIRRTL/FIRRTLTypes.h [hunk 6]
```diff
@@ -292,11 +292,14 @@ class BundleType : public FIRRTLType::TypeBase<BundleType, FIRRTLType,
   /// Look up an element's index by name.  This returns None on failure.
   llvm::Optional<unsigned> getElementIndex(StringRef name);
 
+  /// Look up an element's name by index. This returns None on failure.
+  llvm::Optional<StringRef> getElementName(size_t index);
```
Comment (lattner) [inline include/circt/Dialect/FIRRTL/FIRRTLTypes.h:296]: The only client of this is force unwrapping the optional.  I'd recommend making this assert that the index is valid, and returning a StringRef directly.

**Review by lattner — APPROVED** [review]
This is brilliant!  Please follow this up with a change to make `hw` align with the same model.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 7]
```diff
@@ -1519,40 +1519,30 @@ void ConstantOp::build(OpBuilder &builder, OperationState &result,
...
-            << fieldname.getValue() << "' in bundle type " << inType;
-      return {};
-    }
-    // FIRRTL puts flips on element fields, not on the underlying
-    // types.  The result type of a subfield should strip a flip
-    // if one exists.
-    return elt->type;
-  }
+  auto fieldIndex =
+      getAttr<IntegerAttr>(attrs, "fieldIndex").getValue().getZExtValue();
```
[the remark this answers is not in the document]
Comment (prithayan) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1535]: `inferReturnType` is a static function, we can't use `this->fieldIndex()` here. This is following the same pattern as other operations.

**Comment by prithayan** [comment]
Thanks @lattner , updated with the requested changes. Merging this now.
I will work on updating hw::StructExtractOp also.

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp [hunk 8]
```diff
@@ -775,7 +775,10 @@ static SmallVector<SubfieldOp> getAllFieldAccesses(Value structValue,
   for (auto op : structValue.getUsers()) {
     assert(isa<SubfieldOp>(op));
     auto fieldAccess = cast<SubfieldOp>(op);
-    if (fieldAccess.fieldname() == field) {
+    auto elemIndex =
+        fieldAccess.input().getType().cast<BundleType>().getElementIndex(field);
+    if (elemIndex.hasValue() &&
+        fieldAccess.fieldIndex() == elemIndex.getValue()) {
```
Comment (lattner) [inline lib/Conversion/FIRRTLToHW/LowerToHW.cpp:781]: Can't this `assert(elemIndex.hasValue())`?   The case where this fails is invalid IR right?

File: lib/Dialect/FIRRTL/FIRRTLDialect.cpp [hunk 9]
```diff
@@ -42,8 +42,7 @@ FieldRef circt::firrtl::getFieldRefFromValue(Value value) {
     if (auto subfieldOp = dyn_cast<SubfieldOp>(op)) {
       value = subfieldOp.input();
       auto bundleType = value.getType().cast<BundleType>();
-      auto index =
-          bundleType.getElementIndex(subfieldOp.fieldname()).getValue();
+      auto index = subfieldOp.fieldIndex();
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLDialect.cpp:45]: I think you can just inline this expression now, it is short enough.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (lattner) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:1535]: Got it, thanks!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 10]
```diff
@@ -1519,41 +1519,37 @@ void ConstantOp::build(OpBuilder &builder, OperationState &result,
...
-    mlir::emitError(*loc, "subfield requires bundle operand");
-  return {};
+  // SubfieldOp verifier checks that the field index is valid with number of
+  // subelements.
+  return inType.getElement(fieldIndex).type;
 }
 
 bool SubfieldOp::isFieldFlipped() {
-  auto fieldname = this->fieldname();
+  auto fieldIndex = this->fieldIndex();
```
Comment (lattner) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:1550]: Nice.  I think it would make sense to inline fieldIndex now and drop the `this->`

**Review by lattner — COMMENTED** [review]
Very nice, this is a great step forward!

File: lib/Conversion/FIRRTLToHW/LowerToHW.cpp (hunk 8 above)
Comment (prithayan) [reply lib/Conversion/FIRRTLToHW/LowerToHW.cpp:781]: What if the `field` argument to this `getAllFieldAccesses` is not even an element of the bundle ? Seems like need not be invalid IR, if `getAllFieldAccesses` is called with incorrect string field name.

**Pull request merged**